msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AbstractClientApplicationBase.java

```diff
@@ -545,10 +558,21 @@ public T correlationId(String val) {
                         builder.httpClient)
         );
 
+        //In most cases we will retrieve various auth/token/etc. via instance discovery during a token request,
+        //  the format is well known enough to be stored in the codebase itself, or the instance discovery result is
+        //  provided at app creation so we can cache it and skip discovery.
+        //However, if the authority being set is a generic OIDC authority (rather than Azure AD-style) then we need to perform
+        //  instance discovery when the client app is created, as there are some app-level methods that require the endpoints from
+        //  the instance discovery metadata.
         if (aadAadInstanceDiscoveryResponse != null) {
```

Comment:
It's not a great behavior to fire HTTP requests when creating the "app" object. But maybe add a bug for this to revisit.

Reply:
I agree it's not ideal, however this is the easiest way to do it because of the `authenticationAuthority` field in AbstractApplicationBase which is used to reference the endpoints found in the OIDC discovery: https://github.com/AzureAD/microsoft-authentication-library-for-java/blob/424eea69c795e122171bd6335c253767042519a1/msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AbstractApplicationBase.java#L32

That field is used in a couple dozen places, such as:
- A method called in the confidential client app's constructor: https://github.com/AzureAD/microsoft-authentication-library-for-java/blob/424eea69c795e122171bd6335c253767042519a1/msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ConfidentialClientApplication.java#L94
- A public helper method for creating the URL for an auth code request: https://github.com/AzureAD/microsoft-authentication-library-for-java/blob/424eea69c795e122171bd6335c253767042519a1/msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AbstractClientApplicationBase.java#L173
- Parameter classes for various token request flows

If the OIDC discovery doesn't happen before that field gets referenced, then all of the endpoints will be null. It's definitely something we can revisit and refactor later, but for now we either make the HTTP call in this one place during app creation, or add checks to a bunch of different places to make sure it get initialized.

Reply:
Pls log a bug, it's a P3 as customers have not complained.

Reply:
Just created a task to track this: https://github.com/AzureAD/microsoft-authentication-library-for-java/issues/856